### PR TITLE
Fix hack that remaps env constness.

### PR DESCRIPTION
**Pull request by fee1-dead**
WARNING: might have perf implications.

Are there any more problems with having a constness in the `ParamEnv` now? :)

r? @oli-obk

### CODE REVIEW

File: compiler/rustc_middle/src/ty/mod.rs [hunk 1]
```diff
@@ -575,6 +575,19 @@ impl<'tcx> Predicate<'tcx> {
 
         Some(tcx.mk_predicate(kind))
     }
+
+    pub fn without_const(mut self, tcx: TyCtxt<'tcx>) -> Self {
+        if let PredicateKind::Trait(TraitPredicate { trait_ref, constness, polarity }) = self.kind().skip_binder()
+            && constness != BoundConstness::NotConst
+        {
+            self = tcx.mk_predicate(self.kind().map_bound(|_| PredicateKind::Trait(TraitPredicate {
```
Comment (compiler-errors) [inline compiler/rustc_middle/src/ty/mod.rs:583]: nit:
```suggestion
            self = tcx.mk_predicate(self.kind().rebind(PredicateKind::Trait(TraitPredicate {
```

**Comment by fee1-dead** [comment]
This might just be noise because the benchmarks related to the trait system that I expect could regress didn't (diesel and other benchmarks). Instead all regressions are on the doc mode. @oli-obk can you please double check my analysis and add the triaged label if you agree？ Thanks.

**Comment by oli-obk** [comment]
```
25,703,966  ???:<alloc::vec::Vec<rustc_infer::traits::Obligation<rustc_middle::ty::Predicate>> as alloc::vec::spec_from_iter::SpecFromIter<rustc_infer::traits::Obligation<rustc_middle::ty::Predicate>, core::iter::adapters::map::Map<alloc::vec::into_iter::IntoIter<rustc_infer::traits::Obligation<rustc_middle::ty::Predicate>>, <rustc_trait_selection::traits::select::SelectionContext>::confirm_candidate::{closure
11,779,364  ???:<rustc_trait_selection::traits::select::SelectionContext>::confirm_candidate
 5,913,240  ???:<rustc_middle::ty::Predicate>::without_const
 2,288,209  ???:<rustc_parse::parser::Parser>::parse_path_segment
-2,022,396  ???:<rustc_parse::parser::Parser>::parse_path_start_expr
 1,812,188  ???:<rustc_span::hygiene::ExpnId as rustc_serialize::serialize::Decodable<rustc_metadata::rmeta::decoder::DecodeContext>>::decode
-1,680,053  ???:<rustc_span::hygiene::SyntaxContextData as rustc_serialize::serialize::Decodable<rustc_metadata::rmeta::decoder::DecodeContext>>::decode
   554,031  ???:<rustc_trait_selection::traits::select::SelectionContext>::select
   391,278  ???:<rustc_resolve::Resolver>::create_def
  -322,449  ???:<rustc_resolve::def_collector::DefCollector as rustc_ast::visit::Visitor>::visit_item
  -238,586  ???:<hashbrown::map::HashMap<rustc_middle::ty::sty::Binder<rustc_middle::ty::TraitPredicate>, (), core::hash::BuildHasherDefault<rustc_hash::FxHasher>>>::insert
  -152,955  ???:<alloc::vec::Vec<rustc_infer::traits::Obligation<rustc_middle::ty::Predicate>> as alloc::vec::spec_from_iter::SpecFromIter<rustc_infer::traits::Obligation<rustc_middle::ty::Predicate>, core::iter::adapters::flatten::FlatMap<core::slice::iter::Iter<rustc_middle::ty::Ty>, alloc::vec::Vec<rustc_infer::traits::Obligation<rustc_middle::ty::Predicate>>, <rustc_trait_selection::traits::select::SelectionContext>::collect_predicates_for_types::{closure
   136,401  /build/glibc-sMfBJT/glibc-2.31/string/../sysdeps/x86_64/multiarch/memmove-vec-unaligned-erms.S:__memcpy_sse2_unaligned_erms
  -118,541  ???:<rustc_parse::lexer::StringReader>::next_token
   109,170  ???:<hashbrown::map::HashMap<rustc_span::def_id::DefId, rustc_span::def_id::DefId, core::hash::BuildHasherDefault<rustc_hash::FxHasher>>>::insert
   108,683  ???:_rjem_je_eset_fit
  -106,850  ???:<rustc_middle::traits::specialization_graph::Graph as rustc_trait_selection::traits::specialize::specialization_graph::GraphExt>::insert
   101,581  ???:<rustc_middle::ty::context::CtxtInterners>::intern_predicate
    91,286  ???:rustc_ast::visit::walk_fn::<rustc_ast_passes::feature_gate::PostExpansionVisitor>
   -86,774  ???:_rjem_je_arena_cache_bin_fill_small
    82,536  ???:_rjem_je_tcache_bin_flush_small
   -75,189  ???:<rustc_middle::ty::context::TyCtxt>::intern_substs
   -56,736  ???:rustc_ast::visit::walk_assoc_item::<rustc_ast_passes::feature_gate::PostExpansionVisitor>
   -51,774  ???:<hashbrown::raw::RawTable<(rustc_middle::ty::sty::Binder<rustc_middle::ty::TraitPredicate>, ())>>::insert::<hashbrown::map::make_hasher<rustc_middle::ty::sty::Binder<rustc_middle::ty::TraitPredicate>, rustc_middle::ty::sty::Binder<rustc_middle::ty::TraitPredicate>, (), core::hash::BuildHasherDefault<rustc_hash::FxHasher>>::{closure
    49,484  ???:<rustc_middle::ty::flags::FlagComputation>::add_substs
```

a small rustdoc regression

**Comment by oli-obk** [comment]
This is probably due to rustdoc moving everything through the trait solver multiple times.
The other tiny regressions are just noise:

```
 72,337  ???:_rjem_je_edata_heap_remove_first
 60,510  ???:base_alloc_impl
 39,006  ???:(anonymous namespace)::TwoAddressInstructionPass::runOnMachineFunction(llvm::MachineFunction&)
-38,902  ???:(anonymous namespace)::TwoAddressInstructionPass::runOnMachineFunction(llvm::MachineFunction&) [clone 
 31,543  ???:std::pair<std::_Rb_tree_iterator<std::pair<llvm::MCContext::ELFEntrySizeKey const, unsigned int> >, bool> std::_Rb_tree<llvm::MCContext::ELFEntrySizeKey, std::pair<llvm::MCContext::ELFEntrySizeKey const, unsigned int>, std::_Select1st<std::pair<llvm::MCContext::ELFEntrySizeKey const, unsigned int> >, std::less<llvm::MCContext::ELFEntrySizeKey>, std::allocator<std::pair<llvm::MCContext::ELFEntrySizeKey const, unsigned int> > >::_M_emplace_unique<std::pair<llvm::MCContext::ELFEntrySizeKey, unsigned int> >(std::pair<llvm::MCContext::ELFEntrySizeKey, unsigned int>&&)
-24,616  ???:std::_Rb_tree<llvm::MCContext::ELFEntrySizeKey, std::pair<llvm::MCContext::ELFEntrySizeKey const, unsigned int>, std::_Select1st<std::pair<llvm::MCContext::ELFEntrySizeKey const, unsigned int> >, std::less<llvm::MCContext::ELFEntrySizeKey>, std::allocator<std::pair<llvm::MCContext::ELFEntrySizeKey const, unsigned int> > >::_M_get_insert_unique_pos(llvm::MCContext::ELFEntrySizeKey const&)
-18,557  /build/glibc-sMfBJT/glibc-2.31/string/../sysdeps/x86_64/strcmp.S:strcmp
-15,609  ???:_rjem_je_edata_avail_remove
 10,736  /build/glibc-sMfBJT/glibc-2.31/elf/dl-lookup.c:_dl_lookup_symbol_x
 10,454  ???:llvm::FPPassManager::runOnFunction(llvm::Function&)
```

**Comment by pnkfelix** [comment]
* This regression was anticipated by the PR author and [analyzed](https://github.com/rust-lang/rust/pull/99521#issuecomment-1191511904) by the reviewer.
* marking as triaged.

@rustbot label: +perf-regression-triaged

**Pull request merged**